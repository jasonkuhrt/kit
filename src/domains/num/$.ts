/**
 * Numeric types and utilities with branded types for mathematical constraints.
 *
 * Provides branded number types (Positive, Negative, Even, Odd, etc.) with
 * runtime validation, mathematical operations, range types, and specialized
 * numeric domains like Complex, Ratio, and BigInt. Includes type guards,
 * ordering, and equivalence utilities.
 */
export * as Num from './$$.js'
