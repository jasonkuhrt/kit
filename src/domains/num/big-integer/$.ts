export * as BigInteger from './$$.ts'
