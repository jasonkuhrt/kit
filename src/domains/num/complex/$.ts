export * as Complex from './$$.js'
