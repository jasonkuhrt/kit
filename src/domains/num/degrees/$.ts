// Export the namespace containing all Degrees operations
export * as Degrees from './$$.js'
