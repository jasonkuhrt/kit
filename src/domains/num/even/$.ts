// Export the namespace containing all Even operations
export * as Even from './$$.ts'
