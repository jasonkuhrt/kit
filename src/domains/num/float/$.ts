// Export the namespace containing all Float operations
export * as Float from './$$.js'
