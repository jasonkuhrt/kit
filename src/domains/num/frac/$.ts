// Export the namespace containing all Frac operations
export * as Frac from './$$.js'
