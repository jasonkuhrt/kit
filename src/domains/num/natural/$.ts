// Export the namespace containing all Natural operations
export * as Natural from './$$.js'
