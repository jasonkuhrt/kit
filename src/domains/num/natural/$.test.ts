import { describe, expect, it } from 'vitest'
import { Natural } from './$.js'

describe('isNatural', () => {
  it('returns true for positive integers', () => {
    expect(Natural.is(1)).toBe(true)
    expect(Natural.is(2)).toBe(true)
    expect(Natural.is(100)).toBe(true)
    expect(Natural.is(Number.MAX_SAFE_INTEGER)).toBe(true)
  })

  it('returns false for zero', () => {
    expect(Natural.is(0)).toBe(false)
  })

  it('returns false for negative numbers', () => {
    expect(Natural.is(-1)).toBe(false)
    expect(Natural.is(-100)).toBe(false)
  })

  it('returns false for non-integers', () => {
    expect(Natural.is(1.5)).toBe(false)
    expect(Natural.is(0.1)).toBe(false)
    expect(Natural.is(Math.PI)).toBe(false)
  })

  it('returns false for special values', () => {
    expect(Natural.is(Infinity)).toBe(false)
    expect(Natural.is(-Infinity)).toBe(false)
    expect(Natural.is(NaN)).toBe(false)
  })

  it('returns false for non-numbers', () => {
    expect(Natural.is('1')).toBe(false)
    expect(Natural.is(null)).toBe(false)
    expect(Natural.is(undefined)).toBe(false)
    expect(Natural.is({})).toBe(false)
  })
})

describe('natural', () => {
  it('returns value for positive integers', () => {
    expect(Natural.from(1)).toBe(1)
    expect(Natural.from(100)).toBe(100)
  })

  it('throws for zero', () => {
    expect(() => Natural.from(0)).toThrow('Value must be positive')
  })

  it('throws for negative numbers', () => {
    expect(() => Natural.from(-1)).toThrow('Value must be positive')
  })

  it('throws for non-integers', () => {
    expect(() => Natural.from(1.5)).toThrow('Value must be an integer')
  })
})

describe('tryNatural', () => {
  it('returns value for positive integers', () => {
    expect(Natural.tryFrom(1)).toBe(1)
    expect(Natural.tryFrom(100)).toBe(100)
  })

  it('returns null for invalid values', () => {
    expect(Natural.tryFrom(0)).toBe(null)
    expect(Natural.tryFrom(-1)).toBe(null)
    expect(Natural.tryFrom(1.5)).toBe(null)
  })
})

describe('parseAsNatural', () => {
  it('parses valid natural number strings', () => {
    expect(Natural.parseAsNatural('1')).toBe(1)
    expect(Natural.parseAsNatural('100')).toBe(100)
  })

  it('returns null for invalid strings', () => {
    expect(Natural.parseAsNatural('0')).toBe(null)
    expect(Natural.parseAsNatural('-1')).toBe(null)
    expect(Natural.parseAsNatural('1.5')).toBe(null)
    expect(Natural.parseAsNatural('abc')).toBe(null)
  })
})

describe('nextNatural', () => {
  it('returns next natural for positive numbers', () => {
    expect(Natural.next(5)).toBe(6)
    expect(Natural.next(5.1)).toBe(6)
    expect(Natural.next(5.9)).toBe(6)
  })

  it('returns 1 for non-positive numbers', () => {
    expect(Natural.next(0)).toBe(1)
    expect(Natural.next(-1)).toBe(1)
    expect(Natural.next(-10.5)).toBe(1)
  })
})

describe('prevNatural', () => {
  it('returns previous natural for values > 1', () => {
    expect(Natural.prev(5)).toBe(4)
    expect(Natural.prev(5.9)).toBe(4)
    expect(Natural.prev(2)).toBe(1)
  })

  it('returns null for values <= 1', () => {
    expect(Natural.prev(1)).toBe(null)
    expect(Natural.prev(0.5)).toBe(null)
    expect(Natural.prev(0)).toBe(null)
    expect(Natural.prev(-1)).toBe(null)
  })
})
