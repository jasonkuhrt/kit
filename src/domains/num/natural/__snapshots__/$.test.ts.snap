// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`from > from(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN THROWS
Error: Value must be positive (> 0), got: -1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-100
╠══════════════════════════════════════════════════╣ THEN THROWS
Error: Value must be positive (> 0), got: -100
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN THROWS
Error: Value must be positive (> 0), got: 0
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS
1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1.5
╠══════════════════════════════════════════════════╣ THEN THROWS
Error: Value must be an integer, got: 1.5
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS
100
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(null) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN THROWS
Error: Value must be an integer, got: Infinity
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is("1") 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
"1"
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is({}) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{}
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-100
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(0.1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0.1
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1.5
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2
╠══════════════════════════════════════════════════╣ THEN RETURNS
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(3.141592653589793) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
3.141592653589793
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(9007199254740991) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
9007199254740991
╠══════════════════════════════════════════════════╣ THEN RETURNS
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(null) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(null) 2`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(null) 3`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(null) 4`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(undefined) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
undefined
╠══════════════════════════════════════════════════╣ THEN RETURNS
false
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN RETURNS
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-100
╠══════════════════════════════════════════════════╣ THEN RETURNS
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS
1
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1.5
╠══════════════════════════════════════════════════╣ THEN RETURNS
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS
100
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(null) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN RETURNS
null
╚══════════════════════════════════════════════════╝"
`;
